### PR TITLE
ingest and export region-based single unit classification

### PR DESCRIPTION
**Pull request by ttngu207**

### CODE REVIEW

File: pipeline/ephys.py [hunk 1]
```diff
@@ -708,6 +711,77 @@ class UnitCCF(dj.Computed):
...
+    definition = """
+    -> experiment.Session
+    """
+
+    class UnitClassification(dj.Part):
+        definition = """
+        -> master
+        -> Unit
+        ---
+        classification: enum('good', 'unlabelled')  # single unit classification label ("good"/"unlabelled")
```
Comment (dimitri-yatsenko) [inline pipeline/ephys.py:729]:
```suggestion
        classification: enum('good', 'unlabeled')  # single unit classification label ("good"/"unlabeled")
```

File: pipeline/ephys.py [hunk 2]
```diff
@@ -708,6 +711,77 @@ class UnitCCF(dj.Computed):
...
+        mat_idx = mat['Idx']
+        mat_anno = mat['AnnoName']
+
+        # unit Idx from the .mat file are concatenated unit indexes, not the original unit id
+        # create a `unit_mapping` to map back to the original unit id
+        insertion_numbers, unit_counts = (ProbeInsertion & key).aggr(
+            Unit.proj(), unit_count='count(unit)').fetch('insertion_number', 'unit_count',
+                                                       order_by='insertion_number')
+        unit_count_cumsum = np.concatenate([[0], unit_counts.astype(int).cumsum()[:-1]])
+        unit_count_cumsum = {i: c for i, c in zip(insertion_numbers, unit_count_cumsum)}
```
Comment (dimitri-yatsenko) [inline pipeline/ephys.py:755]:
```suggestion
        unit_count_cumsum = dict(zip(insertion_numbers, unit_count_cumsum))
```

File: pipeline/export/nwb.py [hunk 3]
```diff
@@ -216,14 +219,20 @@ def datajoint_to_nwb(session_key, raw_ephys=False, raw_video=False):
...
         for unit in unit_query.fetch(as_dict=True):
             unit['id'] = max(nwbfile.units.id.data) + 1 if nwbfile.units.id.data else 0
-            aligned_times, trial_numbers, trial_starts, trial_stops = (ephys.Unit.TrialSpikes * experiment.SessionTrial & unit).fetch('spike_times', 'trial', 'start_time', 'stop_time', order_by='trial')
+            aligned_spikes = (ephys.Unit.TrialSpikes & unit).fetch(
+                'spike_times', order_by='trial')
             raw_spike_times = []
-            for aligned_time, trial_number, trial_start, trial_stop in zip(aligned_times, trial_numbers, trial_starts, trial_stops):
-                go_cue_time = (experiment.SessionTrial * experiment.TrialEvent & session_key & {'trial': trial_number} & {'trial_event_type':'go'}).fetch1('trial_event_time', order_by='trial')
-                raw_spike_times.append(aligned_time + float(trial_start) + float(go_cue_time))
+            for aligned_spks, go_cue_time, trial_start, trial_stop in zip(
```
Comment (dimitri-yatsenko) [inline pipeline/export/nwb.py:233]: `trial_stop` is not being used.

File: pipeline/export/nwb.py [hunk 4]
```diff
@@ -216,14 +219,20 @@ def datajoint_to_nwb(session_key, raw_ephys=False, raw_video=False):
...
         unit_query = units_query & insert_key
         for unit in unit_query.fetch(as_dict=True):
             unit['id'] = max(nwbfile.units.id.data) + 1 if nwbfile.units.id.data else 0
-            aligned_times, trial_numbers, trial_starts, trial_stops = (ephys.Unit.TrialSpikes * experiment.SessionTrial & unit).fetch('spike_times', 'trial', 'start_time', 'stop_time', order_by='trial')
+            aligned_spikes = (ephys.Unit.TrialSpikes & unit).fetch(
+                'spike_times', order_by='trial')
             raw_spike_times = []
-            for aligned_time, trial_number, trial_start, trial_stop in zip(aligned_times, trial_numbers, trial_starts, trial_stops):
-                go_cue_time = (experiment.SessionTrial * experiment.TrialEvent & session_key & {'trial': trial_number} & {'trial_event_type':'go'}).fetch1('trial_event_time', order_by='trial')
-                raw_spike_times.append(aligned_time + float(trial_start) + float(go_cue_time))
```
Comment (dimitri-yatsenko) [inline pipeline/export/nwb.py:226]:
```python
raw_spike_times = [
    a + float(g) + float(s)
    for a, g, s in zip(aligned_spikes, go_cue_times, trial_starts)
]

File: pipeline/export/nwb.py [hunk 5]
```diff
@@ -216,14 +219,20 @@ def datajoint_to_nwb(session_key, raw_ephys=False, raw_video=False):
...
+            for aligned_spks, go_cue_time, trial_start, trial_stop in zip(
+                    aligned_spikes, go_cue_times, trial_starts, trial_stops):
+                raw_spike_times.append(aligned_spks + float(trial_start) + float(go_cue_time))
             spikes = np.concatenate(raw_spike_times).ravel()
             observed_times = np.array([trial_starts, trial_stops]).T.astype('float')
             unit['spike_times'] = spikes
```
Comment (dimitri-yatsenko) [inline pipeline/export/nwb.py:238]: There is no reason for the variable `observed_times`. It's just as expressive to directly assign the value.

**Review by dimitri-yatsenko — CHANGES_REQUESTED** [review]

**Review by kushalbakshi — APPROVED** [review]

**Pull request merged**